### PR TITLE
Closes SI-5969.

### PR DESCRIPTION
**Pull request by hubertp**
Assumption that we enter tryTwice when inferring
the right alternative with an empty buffer is
wrong.
In this particular bug it manifested itself on
if/then/else which share the same context and 'else'
branch was simply flushing the buffer with an
error from the 'then' branch.

Review by @retronym, please (you were on the right track judging from the comments in the bug report).

### CODE REVIEW

**Comment by retronym** [comment]
It might be worth encapsulating the save/try/finally/restore pattern into a method in `Context`. That might be awkward here, as you are accessing `saved` in the catch. If the code isn't amenable to that sort of abstraction, could you give it a play-by-play commentary to explain each line?

**Comment by hubertp** [comment]
I moved the restoreState to finally because of CyclicReference error (it should have been from the beginning). Basically other option is to have context.restoreState(saved) before context.hasErrors & in CyclicReference handler. It is a matter of preference which one is better.

I don't think it is necessary to have yet another abstraction in Context for the save pattern as this is pretty specialized for tryTwice in Infer e.g. one could imagine different handlers for TypeErrors/CyclicReferences, also operations that are done in infer often rely on the state of the Context and whether it contains errors. Typically you would be able to handle it nicely using silent(...) but here we are also messing with implicits availability.

So to sum up why we have to do this additional buffer clearing/update because it is impossible to distinguish between errors that occurred before and during infer(...) and our only option is to handle that manually.hth.

**Comment by hubertp** [comment]
PLS REBUILD ALL

**Comment by retronym** [comment]
Thanks for the explanation -- could you please add the last line as a comment in the code? With that, lgtm.

**Comment by hubertp** [comment]
Sure, will do. Thanks for the review.

**Comment by hubertp** [comment]
@adriaanm feel free to merge it, the test suite run cleanly on my test machine. Jenkins seems to be on a personal vendetta against me.

**Comment by adriaanm** [comment]
ok, don't take it personally, though -- it's more like a full-on war against us all
let us unite and we shall prevail!

**Comment by adriaanm** [comment]
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/view/pr-validators/job/pr-scala-testsuite-linux-opt/410/
